### PR TITLE
Bugfix/zcs 12967

### PR DESCRIPTION
**Pull request by zimsuchitgupta**
[Background]
When the mailbox is migrated from other mail product, each message's unique-id will be changed because Zimbra's POP server generates its own unique-id for the UDIL command - "[msg-id].[digest]". As a result, after the migration, the end user who is using POP3 client may have to download all the messages again (because a new unique-id is assigned again).

This enhancement introduce a new sub-field in METADATA field in the mail_item table to store a custom unique-id string, and return the value of sub-field as a custom unique-id text string when the end user executes the POP3 UIDL command.

[New SOAP API]

({uid})* ()* [New attribute] zimbraFeatureCustomUIDEnabled type: boolean optionalIn: account,cos,domain description: Whether the custom UID is used when the POP3 UIDL command is submitted. Set TRUE to use custom UID string which is set by SOAP request SetPop3UIDRequest. If any custom UID string is not set or the value is an empty string, the standard Zimbra UID is used. Set FALSE to use a standard Zimbra UID ("id.digest"). (Note) This attribute affects only the behaviour of the POP3 UIDL command. It does not affect to the return value of GetPop3UIDRequest API
[New METADATA sub-field]
name: p3uid
value type: string
(note) This field does not appear in the response of the existing APIs, such as GetMsgMetadataRequest, GetCustomMetadataRequest, and GetMailboxMetadataRequest.

### CODE REVIEW

**Review by vishnukantjangid — APPROVED** [review]

File: store/src/java/com/zimbra/cs/mailbox/MailItem.java [hunk 1]
```diff
@@ -776,6 +776,11 @@ public long guessSize() {
     protected CustomMetadataList mExtendedData;
     protected ACL                rights;
 
+    /**
+     * value stored in the Metadata.FN_POP3_UID ("p3uid")
+     */
+    protected String             mPop3Uid;
```
Comment (shrutig0510) [inline store/src/java/com/zimbra/cs/mailbox/MailItem.java]: Formatting issue

File: soap/src/java/com/zimbra/soap/mail/message/GetPop3UIDRequest.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
+/*
+ * ***** BEGIN LICENSE BLOCK *****
+ * Zimbra Collaboration Suite Server
+ * Copyright (C) 2019 Synacor, Inc.
```
Comment (k-kato) [inline soap/src/java/com/zimbra/soap/mail/message/GetPop3UIDRequest.java]: Check other files as well.
```suggestion
 * Copyright (C) 2023 Synacor, Inc.
```

File: common/src/java/com/zimbra/common/soap/MailConstants.java [hunk 3]
```diff
@@ -305,6 +305,11 @@ private MailConstants() {
     public static final String E_FILE_SHARED_WITH_ME_REQUEST = "FileSharedWithMeRequest";
     public static final String E_FILE_SHARED_WITH_ME_RESPONSE = "FileSharedWithMeResponse";
 
+	public static final String E_GET_POP3_UID_REQUEST = "GetPop3UIDRequest";
+	public static final String E_GET_POP3_UID_RESPONSE = "GetPop3UIDResponse";
+	public static final String E_SET_POP3_UID_REQUEST = "SetPop3UIDRequest";
+	public static final String E_SET_POP3_UID_RESPONSE = "SetPop3UIDResponse";
+
```
Comment (k-kato) [inline common/src/java/com/zimbra/common/soap/MailConstants.java]: indentation

File: common/src/java/com/zimbra/common/soap/MailConstants.java [hunk 4]
```diff
@@ -611,6 +616,12 @@ private MailConstants() {
     public static final QName FILE_SHARED_WITH_ME_REQUEST = QName.get(E_FILE_SHARED_WITH_ME_REQUEST, NAMESPACE);
     public static final QName FILE_SHARED_WITH_ME_RESPONSE = QName.get(E_FILE_SHARED_WITH_ME_RESPONSE, NAMESPACE);
 
+    // POP3 UIDL
+	public static final QName GET_POP3_UID_REQUEST = QName.get(E_GET_POP3_UID_REQUEST, NAMESPACE);
+	public static final QName GET_POP3_UID_RESPONSE = QName.get(E_GET_POP3_UID_RESPONSE, NAMESPACE);
+	public static final QName SET_POP3_UID_REQUEST = QName.get(E_SET_POP3_UID_REQUEST, NAMESPACE);
+	public static final QName SET_POP3_UID_RESPONSE = QName.get(E_SET_POP3_UID_RESPONSE, NAMESPACE);
+
```
Comment (k-kato) [inline common/src/java/com/zimbra/common/soap/MailConstants.java]: indentation

File: store/src/java/com/zimbra/cs/mailbox/MailItem.java (hunk 1 above)
Comment (zimsuchitgupta) [reply store/src/java/com/zimbra/cs/mailbox/MailItem.java]: Done

File: soap/src/java/com/zimbra/soap/mail/message/GetPop3UIDRequest.java (hunk 2 above)
Comment (zimsuchitgupta) [reply soap/src/java/com/zimbra/soap/mail/message/GetPop3UIDRequest.java]: Done

File: common/src/java/com/zimbra/common/soap/MailConstants.java (hunk 3 above)
Comment (zimsuchitgupta) [reply common/src/java/com/zimbra/common/soap/MailConstants.java]: Done

File: common/src/java/com/zimbra/common/soap/MailConstants.java (hunk 4 above)
Comment (zimsuchitgupta) [reply common/src/java/com/zimbra/common/soap/MailConstants.java]: Done

File: soap/src/java/com/zimbra/soap/mail/message/SetPop3UIDResponse.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
+/*
+ * ***** BEGIN LICENSE BLOCK *****
+ * Zimbra Collaboration Suite Server
+ * Copyright (C) 2019 Synacor, Inc.
```
Comment (k-kato) [inline soap/src/java/com/zimbra/soap/mail/message/SetPop3UIDResponse.java]:
```suggestion
 * Copyright (C) 2023 Synacor, Inc.
```

File: store/src/java/com/zimbra/cs/db/DbMailItem.java [hunk 6]
```diff
@@ -1,7 +1,7 @@
 /*
  * ***** BEGIN LICENSE BLOCK *****
  * Zimbra Collaboration Suite Server
- * Copyright (C) 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016 Synacor, Inc.
+ * Copyright (C) 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2019, 2023 Synacor, Inc.
```
Comment (k-kato) [inline store/src/java/com/zimbra/cs/db/DbMailItem.java]:
```suggestion
 * Copyright (C) 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2023 Synacor, Inc.
```

File: store/src/java/com/zimbra/cs/pop3/Pop3Handler.java [hunk 7]
```diff
@@ -1,7 +1,7 @@
 /*
  * ***** BEGIN LICENSE BLOCK *****
  * Zimbra Collaboration Suite Server
- * Copyright (C) 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016 Synacor, Inc.
+ * Copyright (C) 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2019, 2023 Synacor, Inc.
```
Comment (k-kato) [inline store/src/java/com/zimbra/cs/pop3/Pop3Handler.java]:
```suggestion
 * Copyright (C) 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2023 Synacor, Inc.
```

**Review by k-kato — APPROVED** [review]

File: store/src/java/com/zimbra/cs/pop3/Pop3Message.java [hunk 8]
```diff
@@ -71,4 +87,17 @@ boolean isDeleted() {
...
+            // Classic UID string
+            return id + "." + digest;
+        } else {
+            // Custom UID stored in the metadata
+            return uid;
+        }
+    }
+    void setUid(String uid) {
+		this.uid = uid;
+	}
```
Comment (k-kato) [inline store/src/java/com/zimbra/cs/pop3/Pop3Message.java]: indentation

File: soap/src/java/com/zimbra/soap/mail/message/SetPop3UIDResponse.java (hunk 5 above)
Comment (zimsuchitgupta) [reply soap/src/java/com/zimbra/soap/mail/message/SetPop3UIDResponse.java]: Done

File: store/src/java/com/zimbra/cs/db/DbMailItem.java (hunk 6 above)
Comment (zimsuchitgupta) [reply store/src/java/com/zimbra/cs/db/DbMailItem.java]: done

File: store/src/java/com/zimbra/cs/pop3/Pop3Handler.java (hunk 7 above)
Comment (zimsuchitgupta) [reply store/src/java/com/zimbra/cs/pop3/Pop3Handler.java]: done

File: store/src/java/com/zimbra/cs/pop3/Pop3Message.java (hunk 8 above)
Comment (zimsuchitgupta) [reply store/src/java/com/zimbra/cs/pop3/Pop3Message.java]: done

Comment (k-kato) [reply store/src/java/com/zimbra/cs/pop3/Pop3Message.java]: it has not been fixed yet... check lines 101-103

Comment (zimsuchitgupta) [reply store/src/java/com/zimbra/cs/pop3/Pop3Message.java]: Plz check now.

**Review by preshitaagrawal — APPROVED** [review]

File: store/src/java/com/zimbra/cs/pop3/Pop3Message.java (hunk 8 above)
Comment (k-kato) [reply store/src/java/com/zimbra/cs/pop3/Pop3Message.java]: confirmed it has been fixed. Thanks!

**Review by vishnukantjangid — APPROVED** [review]

**Pull request merged**